README.md
# pkgtop
Show largest installed packages.

## Installation
```bash
curl https://raw.githubusercontent.com/SharUpOff/pkgtop/main/dist/pkgtop.sh -so - | sudo tee /usr/local/bin/pkgtop > /dev/null
```
```bash
sudo chmod 755 /usr/local/bin/pkgtop
```

## Usage
```bash
pkgtop
```
Output:
```
libc6..............................................................   13.27 MiB 
perl-base..........................................................    7.59 MiB 
coreutils..........................................................    6.95 MiB 
dpkg...............................................................    6.58 MiB 
libssl3............................................................    5.69 MiB 
apt................................................................    4.06 MiB 
util-linux.........................................................    3.32 MiB 
libapt-pkg6.0......................................................    3.10 MiB 
libstdc++6.........................................................    2.69 MiB 
libc-bin...........................................................    2.48 MiB 
```

The colour output makes it possible to compare installed packages visually:

![screenshot](pkgtop.png)

## Arguments
```bash
$ pkgtop [lines [columns]] [--skip <count>] [--exclude <name>] [--mark <name>] [--other] [--total] [--all] [--raw] [--version] [--help]
```

## Features
- Follow terminal and prompt dimensions by default;
- Specify lines:
  ```bash
  $ pkgtop 5
  libc6..............................................................   13.27 MiB 
  perl-base..........................................................    7.59 MiB 
  coreutils..........................................................    6.95 MiB 
  dpkg...............................................................    6.58 MiB 
  libssl3............................................................    5.69 MiB 
  ```
- Specify columns:
  ```bash
  $ pkgtop 5 42
  libc6........................   13.27 MiB 
  perl-base....................    7.59 MiB 
  coreutils....................    6.95 MiB 
  dpkg.........................    6.58 MiB 
  libssl3......................    5.69 MiB 
  ```
- Show other:
  ```bash
  $ pkgtop 5 --other
  [other]............................................................   61.35 MiB 
  libc6..............................................................   13.27 MiB 
  perl-base..........................................................    7.59 MiB 
  coreutils..........................................................    6.95 MiB 
  dpkg...............................................................    6.58 MiB 
  ```
- Show total:
  ```bash
  $ pkgtop 5 --other --total
  [other]............................................................   67.93 MiB 
  libc6..............................................................   13.27 MiB 
  perl-base..........................................................    7.59 MiB 
  coreutils..........................................................    6.95 MiB 
  [total]............................................................   95.74 MiB 
  ```
- Skip packages:
  ```bash
  $ pkgtop 5 --other --total --skip 2
  [other]............................................................   55.67 MiB 
  coreutils..........................................................    6.95 MiB 
  dpkg...............................................................    6.58 MiB 
  libssl3............................................................    5.69 MiB 
  [total]............................................................   74.87 MiB 
  ```
- Exclude package:
  ```bash
  $ pkgtop 5 --other --total --exclude libc6
  [other]............................................................   61.35 MiB 
  perl-base..........................................................    7.59 MiB 
  coreutils..........................................................    6.95 MiB 
  dpkg...............................................................    6.58 MiB 
  [total]............................................................   82.46 MiB 
  ```
- Exclude multiple packages:
  ```bash
  $ pkgtop 5 --other --total --exclude libc6 --exclude coreutils
  [other]............................................................   55.67 MiB 
  perl-base..........................................................    7.59 MiB 
  dpkg...............................................................    6.58 MiB 
  libssl3............................................................    5.69 MiB 
  [total]............................................................   75.52 MiB 
  ```
- Mark package:
  ```bash
  $ pkgtop 5 --other --total --mark coreutils
  [other]............................................................   67.93 MiB 
  libc6..............................................................   13.27 MiB 
  perl-base..........................................................    7.59 MiB 
  coreutils..........................................................    6.95 MiB<
  [total]............................................................   95.74 MiB 
  ```
- Mark multiple packages:
  ```bash
  $ pkgtop  5 --other --total --mark coreutils --mark [other]
  [other]............................................................   67.93 MiB<
  libc6..............................................................   13.27 MiB 
  perl-base..........................................................    7.59 MiB 
  coreutils..........................................................    6.95 MiB<
  [total]............................................................   95.74 MiB
  ```
- Do not limit the output:
  ```bash
  $ pkgtop --all
  libc6..............................................................   13.27 MiB 
  perl-base..........................................................    7.59 MiB
  ...
  libtirpc-common....................................................   32.00 KiB 
  libaudit-common....................................................   23.00 KiB 
  ```

### macOS
> :warning: Use `zsh` on `macOS`.
> This script uses associative arrays introduced in `bash` **>= 4**, while `macOS` have only **3.X.X**.

```zsh
zsh pkgtop.sh
```

# Dependencies

## ArchLinux
You may want to install `expac` to improve script performance:
```bash
$ time pkgtop &> /dev/null

real    0m0,576s
user    0m0,527s
sys     0m0,067s

$ sudo pacman -S expac
$ time pkgtop &> /dev/null

real    0m0,391s
user    0m0,245s
sys     0m0,120s
```

# Compatibility
- [x] GNU/Linux
  - [x] Ubuntu
    - [x] 22.04 (TEST OK: 2023-09-10)
    - [x] 20.04 (TEST OK: 2023-09-10)
    - [x] 18.04 (TEST OK: 2023-09-10)
    - [x] 16.04 (TEST OK: 2023-09-10)
    - [x] 14.04 (TEST OK: 2023-09-10)
  - [x] Debian
    - [x] 11 (TEST OK: 2023-09-10)
    - [x] 10 (TEST OK: 2023-09-10)
    - [x] 9 (TEST OK: 2023-09-10)
  - [x] Fedora
    - [x] 37 (TEST OK: 2023-09-10)
    - [x] 36 (TEST OK: 2023-09-10)
    - [x] 35 (TEST OK: 2023-09-10)
    - [x] 34 (TEST OK: 2023-09-10)
  - [x] RedHat
    - [x] 9 (TEST OK: 2023-09-10)
    - [x] 8 (TEST OK: 2023-09-10)
  - [x] CentOS
    - [x] 7 (TEST OK: 2023-09-10)
  - [x] OpenSUSE
    - [x] tumbleweed (TEST OK: 2023-09-10)
    - [x] leap
      - [x] 15 (TEST OK: 2023-09-10)
  - [x] ArchLinux (TEST OK: 2023-09-10)
  - [x] OpenWRT (TEST OK: 2022-07-16)
- [x] Other operating systems
  - [x] macOS
    - [x] 13 (TEST OK: 2023-09-09)
- [x] Multiplatform package management systems
  - [x] Homebrew (TEST OK: 2023-09-09)

# Contribution
🛠 You are welcome to add support for other distributions, fix bugs or improve functionality. Please, do not forget to add tests.

## Add distribution support
Create a plugin file `src/includes/distributions/yourdistro/01_package-manager.sh`
```bash
# YourDistro (package-manager)
if command -v package-manager &> /dev/null; then
    # write installed packages to the STDOUT using format: %{bytes}d %{name}s\n
    package-manager --installed --format='%{bytes}d %{name}s\n'

    # prevent other plugins from running
    exit $?  
fi
```

## Add multiplatform package management system support
Create a plugin file `src/includes/multiplatform/yourpackagesystem/01_package-manager.sh`
```bash
# YourPackageSystem (package-manager)
if command -v package-manager &> /dev/null; then
    # write installed packages to the STDOUT using format: %{bytes}d %{name}s\n
    package-manager --installed --format='%{bytes}d %{name}s\n'

    # other plugins can also be run to combine the results
fi
```

Yor plugin should write installed packages using format `%{bytes}d %{name}s\n` to the `STDOUT`:
```bash
2131 foo
34534 bar
```

# Testing
There are some regression tests provided for compatible distributions.

## Run tests
```bash
docker compose -f docker-tests-run.yml up
```
Output:
```bash
pkgtop-test-ubuntu-22.04-1 exited with code 0
pkgtop-test-ubuntu-20.04-1 exited with code 0
...
```

## How does testing work
- The actual script is running in a docker container for each of compatible distributions;
- The current terminal output is compared with previously created reference file:
```bash
bash ./dist/pkgtop.sh 25 80 | diff ./data/ubuntu-22.04.txt -
```

## Update reference data
```bash
docker compose -f docker-tests-update.yml up
```


CHANGELOG.md
# Changelog

## 1.1.0

### Features
- Support for other operating systems
  - macOS
- Support for multiplatform package management systems
  - Homebrew

### Fixes
- Compatibility
  - zsh
  - awk without unicode support

## 1.0.0

### Features
- Shows largest installed packages
- Follows terminal and prompt dimensions by default
- Colour output by default
- Support for custom output size
- Options to show `other` and `total` packages size
- Options to `skip` first N and `exclude` specified packages
- Option to `mark` specified packages
- Option for unlimited output
- Support for GNU/Linux distributions
  - dpkg
    - Ubuntu (>=14.04, <=22.04)
    - Debian (>=9, <=11)
  - rpm
    - Fedora (>=34, <=37)
    - RedHat (>=8, <=9)
    - CentOS (==7)
    - OpenSUSE (tumbleweed, leap==15)
  - pacman (expac can be used for better performance)
    - ArchLinux
  - opkg
    - OpenWRT


src/includes/parse_arguments.sh
# default lines and columns
DEFAULT_LINES=25
DEFAULT_COLUMNS=80
LIMIT_COLUMNS=80

# the settings array is used to store the values parsed from the command line arguments.
declare -A settings

# set default settings
settings[exclude]=""
settings[other]=0
settings[total]=0
settings[skip]=0

# detect tty
test ! -t 1
settings[tty]=$?

# process positional arguments: [lines [columns]]
parse_number()
{
    ##################### the input string:
    #
    #      +------------- can contain one minus sign at the beginning;
    #      |      +------ the second character should be a digit 1 to 9;
    #      |      |    +- can only contain digits 0 to 9 until the end;
    #      |      |    |
    egrep '^-{0,1}[1-9][0-9]*$' <<< "${1}"

    return $?
}

# set the lines option if the first argument is a number
settings[lines]="$(parse_number "${1}")"

# process the second argument only if lines option is set
if [ ! -z "${settings[lines]}" ]; then
    # set the columns option if the second argument is a number
    settings[columns]="$(parse_number "${2}")"
fi

# process keyword arguments
for arg in $@; do

    # an option context may be declared by the previous argument,
    # in which case the current argument is the value for that option
    if [ ! -z "${context[name]}" ]; then
        # mark the option context to be closed after processing the current argument
        context[close]=1
    fi

    case $arg in
        --help|-h)
            echo -n "Usage: ${0} [lines [columns]] "
            echo "[--skip <count>] [--exclude <name>] [--mark <name>] "
            echo "[--other] [--total] [--all] [--raw] [--version] [--help]"
            echo
            echo "  [lines] --lines <lines> --lines=<lines> -l <lines>"
            echo "    Show specified number of lines (results)."
            echo "    If [other] and/or [total] lines are enabled they are also included in the number of lines."
            echo
            echo "  [columns] --columns <columns> --columns=<columns> -c <columns>"
            echo "    Show specified number of columns (width)."
            echo
            echo "  --other --show-other -o"
            echo "    Show the total size of packages not included in the list of results."
            echo "    The result named [other] is sorted along with other results."
            echo "    <!> Excluded packages are ignored in the count."
            echo
            echo "  --total --show-total -t"
            echo "    Show the total size of all packages."
            echo "    The result named [total] is displayed at the end of the list."
            echo "    <!> Excluded packages are ignored in the count."
            echo
            echo "  --all --show-all -a"
            echo "    Do not limit the output. Display all packages instead."
            echo "    <!> Excluded packages stay hidden even if all results should be displayed."
            echo
            echo "  --raw -r"
            echo "    Preserve colour output even if tty is not detected."
            echo
            echo "  --skip <count> --skip=<count> -s <count>"
            echo "    Skip specified number of first package(s)."
            echo "    The argument can be specified multiple times."
            echo "    <!> Skipped packages do not count towards [total] or [other] results."
            echo "    <!> Skipped packages stay hidden even if all results should be displayed."
            echo
            echo "  --exclude <name> --exclude=<name> -e <name>"
            echo "    Exclude specified package(s)."
            echo "    The argument can be specified multiple times."
            echo "    <!> Excluded packages do not count towards [total] or [other] results."
            echo "    <!> Excluded packages stay hidden even if all results should be displayed."
            echo
            echo "  --mark <name> --mark=<name> -m <name>"
            echo "    Mark specified package(s)."
            echo
            echo "  --version -v"
            echo "    Show version info and exit."
            echo
            echo "  --help -h"
            echo "    Show this info and exit."
            echo
            exit 0
        ;;
        --version|-v)
            echo "$(cat ../VERSION)"
            exit 0
        ;;
        --other|--show-other|-o)
            settings[other]=1
        ;;
        --total|--show-total|-t)
            settings[total]=1
        ;;
        --all|--show-all|-a)
            settings[lines]=-1;
        ;;
        --raw|-r)
            settings[tty]=1;
        ;;
        --lines|-l)
            declare -A context
            context[name]='lines'
        ;;
        --lines=*)
            settings[lines]="${arg/--lines=}"
        ;;
        --columns|-c)
            declare -A context
            context[name]='columns'
        ;;
        --columns=*)
            settings[columns]="${arg/--columns=}"
        ;;
        --skip|-s)
            declare -A context
            context[name]='skip'
        ;;
        --skip=*)
            settings[skip]="${arg/--skip=}"
        ;;
        --exclude|-e)
            declare -A context
            context[name]='exclude'
            context[multiple]=1
        ;;
        --exclude=*)
            settings[exclude]="${settings[exclude]} ${arg/--exclude=}"
        ;;
        --mark|-m)
            declare -A context
            context[name]='mark'
            context[multiple]=1
        ;;
        --mark=*)
            settings[mark]="${settings[mark]} ${arg/--mark=}"
        ;;
        *)
            # an option context may be declared by the previous argument,
            # in which case the current argument is the value for that option
            if [ ! -z "${context[name]}" ]; then
                # the option can take multiple values if the multiple flag is set in the context
                if [ ! -z "${context[multiple]}" ]; then
                    settings[${context[name]}]="${settings[${context[name]}]} ${arg}"
                else
                    settings[${context[name]}]="${arg}"
                fi
            fi
        ;;
    esac

    # an option context may be declared by the previous argument,
    # in which case the current argument is the value for that option
    if [ ! -z "${context[close]}" ]; then
        # now the time to close the option context
        unset context
    fi
done
